File: ontrack-extension-jenkins/src/main/java/net/nemerosa/ontrack/extension/jenkins/JenkinsJobDecorationExtension.java
package net.nemerosa.ontrack.extension.jenkins;

import net.nemerosa.ontrack.extension.api.DecorationExtension;
import net.nemerosa.ontrack.extension.jenkins.client.JenkinsClient;
import net.nemerosa.ontrack.extension.jenkins.client.JenkinsClientFactory;
import net.nemerosa.ontrack.extension.jenkins.client.JenkinsJob;
import net.nemerosa.ontrack.extension.support.AbstractExtension;
import net.nemerosa.ontrack.model.structure.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.EnumSet;
import java.util.List;

/**
 * Provides a decoration that displays the state of a running job.
 */
@Component
public class JenkinsJobDecorationExtension extends AbstractExtension implements DecorationExtension<JenkinsJob> {

    private final PropertyService propertyService;
    private final JenkinsClientFactory jenkinsClientFactory;

    @Autowired
    public JenkinsJobDecorationExtension(JenkinsExtensionFeature extensionFeature, PropertyService propertyService, JenkinsClientFactory jenkinsClientFactory) {
        super(extensionFeature);
        this.propertyService = propertyService;
        this.jenkinsClientFactory = jenkinsClientFactory;
    }

    @Override
    public EnumSet<ProjectEntityType> getScope() {
        return EnumSet.of(
                ProjectEntityType.PROJECT,
                ProjectEntityType.BRANCH,
                ProjectEntityType.PROMOTION_LEVEL,
                ProjectEntityType.VALIDATION_STAMP
        );
    }

    @Override
    public List<Decoration<JenkinsJob>> getDecorations(ProjectEntity entity) {
        // Gets the Jenkins Job property for this entity, if any
        Property<JenkinsJobProperty> property = propertyService.getProperty(entity, JenkinsJobPropertyType.class.getName());
        if (property.isEmpty()) {
            return Collections.emptyList();
        } else {
            // Gets a client
            // FIXME getJob does not need a full HTTP client
            JenkinsClient jenkinsClient = jenkinsClientFactory.getClient(property.getValue().getConfiguration());
            // Gets the Jenkins job
            JenkinsJob job = jenkinsClient.getJob(property.getValue().getJob());
            // Gets the decoration for the job
            return Collections.singletonList(
                    getDecoration(job)
            );
        }
    }

    private Decoration<JenkinsJob> getDecoration(JenkinsJob job) {
        return Decoration.of(this, job);
    }

}

File: ontrack-extension-jenkins/src/main/java/net/nemerosa/ontrack/extension/jenkins/JenkinsJobDecorationExtension.kt
package net.nemerosa.ontrack.extension.jenkins

import net.nemerosa.ontrack.extension.api.DecorationExtension
import net.nemerosa.ontrack.extension.jenkins.client.JenkinsClientFactory
import net.nemerosa.ontrack.extension.jenkins.client.JenkinsJob
import net.nemerosa.ontrack.extension.support.AbstractExtension
import net.nemerosa.ontrack.model.structure.Decoration
import net.nemerosa.ontrack.model.structure.ProjectEntity
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import net.nemerosa.ontrack.model.structure.PropertyService
import org.springframework.stereotype.Component
import java.util.*

/**
 * Provides a decoration that displays the state of a running job.
 */
@Component
class JenkinsJobDecorationExtension(
    extensionFeature: JenkinsExtensionFeature,
    private val propertyService: PropertyService,
    private val jenkinsClientFactory: JenkinsClientFactory
) : AbstractExtension(extensionFeature), DecorationExtension<JenkinsJob> {

    override fun getScope(): EnumSet<ProjectEntityType> {
        return EnumSet.of(
            ProjectEntityType.PROJECT,
            ProjectEntityType.BRANCH,
            ProjectEntityType.PROMOTION_LEVEL,
            ProjectEntityType.VALIDATION_STAMP
        )
    }

    override fun getDecorations(entity: ProjectEntity): List<Decoration<JenkinsJob>> {
        // Gets the Jenkins Job property for this entity, if any
        return propertyService.getPropertyValue(entity, JenkinsJobPropertyType::class.java)
            ?.let {
                // Gets a client
                val jenkinsClient = jenkinsClientFactory.getClient(it.configuration)
                // Gets the Jenkins job
                val job = jenkinsClient.getJob(it.job)
                // Gets the decoration for the job
                listOf(
                    getDecoration(job)
                )
            }
            ?: emptyList()
    }

    private fun getDecoration(job: JenkinsJob): Decoration<JenkinsJob> {
        return Decoration.of(this, job)
    }
}
